chi2/chinotype.py

```diff
@@ -355,9 +369,28 @@ def prepChi(self):
                 select distinct obs.patient_num pn, concept_cd ccd
                 from {1}.observation_fact obs
                 join {2} chipat on chipat.pn = obs.patient_num
-                '''.format(pconcepts, schema, self.chipats)
+                union all
+                select distinct obs.patient_num pn, c_basecode ccd 
+                from {3}.{4} 
+                left join {1}.concept_dimension cd 
+                on concept_path like c_dimcode||'%' 
+                left join {1}.observation_fact obs 
+                on cd.concept_cd = obs.concept_cd 
+                join {2} chipat on chipat.pn = obs.patient_num
+                -- selection criteria for specific types of branch nodes
+                where ( {5} ) and 
+                -- selection criteria affecting all branch nodes
+                {6}
+                '''.format(pconcepts, schema, self.chipats, self.metaschema, self.termtable, self.branchnodes, self.allbranchnodes)
+                cols, rows = do_log_sql(db, sql)
+                sql = '''
+                create index {0}_pn_idx on {0} (pn)
+                '''.format(pconcepts)
+                cols, rows = do_log_sql(db, sql)
+                sql = '''
+                create index {0}_ccd_idx on {0} (ccd)
+                '''.format(pconcepts)
```

Comment:
From line 372 to here is the heart of the changes associated with the rollup itself. We're simply adding additional rows of counts aggregated by various branch/folder nodes to what the original `chi_pconcepts` table would have been (which ere counts aggregated only by leaf nodes).